src/features/Send/useCases/StartMultiTokenTx/InputReceiver/ResolveAddress.tsx

```diff
@@ -0,0 +1,119 @@
+import {Resolution} from '@unstoppabledomains/resolution'
+import _ from 'lodash'
+import React from 'react'
+import {Text, View, ViewProps} from 'react-native'
+import {useQuery, UseQueryOptions} from 'react-query'
+
+import {HelperText} from '../../../../../components'
+import {getNetworkConfigById, NetworkId, normalizeToAddress, YoroiWallet} from '../../../../../yoroi-wallets'
+import {useStrings} from '../../../common/strings'
+import {InputReceiver} from './InputReceiver'
+
+type ReceiverProps = ViewProps & {
+  receiver: string
+  address: string
+  errorMessage: string
+  isLoading: boolean
+  onChangeReceiver: (receiver: string) => void
+}
+export const ResolveAddress = ({
+  isLoading,
+  address,
+  receiver,
+  errorMessage,
+  onChangeReceiver,
+  style,
+  ...props
+}: ReceiverProps) => {
+  const strings = useStrings()
+  const isResolved = !isLoading && isDomain(receiver) && !receiver.includes(address)
+  const isError = errorMessage.length > 0
+
+  return (
+    <View style={style} {...props}>
+      <InputReceiver
+        value={receiver}
+        onChangeText={onChangeReceiver}
+        error={isError}
+        errorText={errorMessage}
+        isLoading={isLoading}
+        autoComplete={false}
+      />
+
+      <HelperText type={isError ? 'error' : 'info'}>
+        {isLoading ? <Text>{strings.pleaseWait}</Text> : <Text>{errorMessage}</Text>}
+      </HelperText>
+
+      {isResolved && (
+        <Text ellipsizeMode="middle" numberOfLines={1}>
+          {`${strings.resolvesTo}: ${address}`}
+        </Text>
+      )}
+    </View>
+  )
+}
+
+export const useReceiver = (
+  {wallet, receiver}: {wallet: YoroiWallet; receiver: string},
+  options?: UseQueryOptions<string, Error, string, [string, 'receiver']>,
+) => {
+  const query = useQuery({
+    queryKey: [receiver, 'receiver'],
+    queryFn: () => resolveAndCheckAddress(receiver, wallet.networkId),
+    ...options,
+  })
+
+  return {
+    ...query,
+    address: query.data ?? '',
+  }
+}
+
+const resolveAndCheckAddress = async (receiver: string, networkId: NetworkId) => {
+  let address = receiver
+  if (isDomain(receiver)) {
+    address = await getUnstoppableDomainAddress(receiver)
+  }
+
+  await isReceiverAddressValid(address, networkId)
+  return address
+}
+
+export const getAddressErrorMessage = (error: Error & {code?: string}, strings: ReturnType<typeof useStrings>) => {
+  switch (error?.code) {
+    case 'UnsupportedDomain':
+      return strings.domainUnsupportedError
+    case 'RecordNotFound':
+      return strings.domainRecordNotFoundError
+    case 'UnregisteredDomain':
+      return strings.domainNotRegisteredError
+    default:
+      return strings.addressInputErrorInvalidAddress
+  }
+}
+
+const getUnstoppableDomainAddress = (receiver: string) => {
+  const resolution = new Resolution()
+  return resolution.addr(receiver, 'ADA')
+}
+
+const isReceiverAddressValid = async (resolvedAddress: string, walletNetworkId: NetworkId): Promise<void> => {
```

Comment:
im not a fan of `Promise<void>` for things that arent mutations

Reply:
yap, this is the issue with the CSL, everything is async, and resolving to the address that was passed looks awkward

Reply:
I mean, I would prefer Promise<boolean>